network-ntt/src/client.rs

```diff
@@ -0,0 +1,297 @@
+use network_core::client::{self as core_client, block::BlockService};
+use protocol::{
+    network_transport::LightWeightConnectionId, protocol::BlockHeaders, protocol::GetBlockHeaders,
+    protocol::GetBlocks, Inbound, Message, Response,
+};
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::net::SocketAddr;
+use tokio::net::TcpStream;
+use tokio::prelude::*;
+
+/// A handle that can be used in order for communication
+/// with the client thread.
+pub struct ClientHandle<T: Block, Tx> {
+    channel: mpsc::UnboundedSender<Request<T>>,
+    phantom: PhantomData<Tx>,
+}
+
+/// Connect to the remote client. Returns future that can
+/// be run on any executor.
+pub fn connect<B, Tx>(
+    sockaddr: SocketAddr,
+) -> impl Future<
+    Item = (impl Future<Item = (), Error = ()>, ClientHandle<B, Tx>),
+    Error = core_client::Error,
+>
+where
+    B: Block + HasHeader,
+    B: cbor_event::Deserialize + cbor_event::Serialize,
+    <B as Block>::Id: cbor_event::Deserialize + cbor_event::Serialize,
+    B::Header: cbor_event::Deserialize + cbor_event::Serialize,
+    B::Header: Header<Id = <B as Block>::Id, Date = <B as Block>::Date>,
+    B: core::fmt::Debug,
+    <B as Block>::Id: core::fmt::Debug,
+    <B as Block>::Date: core::fmt::Debug,
+    Tx: cbor_event::Serialize + cbor_event::Deserialize,
+    Tx: TransactionId,
+{
+    TcpStream::connect(&sockaddr)
+        .map_err(move |err| core_client::Error::new(core_client::ErrorKind::Rpc, err))
+        .and_then(move |stream| {
+            protocol::Connection::connect(stream)
+                .map_err(move |err| core_client::Error::new(core_client::ErrorKind::Rpc, err))
+                .and_then(move |connection: protocol_tokio::Connection<_, B, Tx>| {
+                    let (stream, source) = mpsc::unbounded();
+                    let handle = ClientHandle {
+                        channel: stream,
+                        phantom: PhantomData,
+                    };
+                    future::ok((run_connection(connection, source), handle))
+                })
+        })
+}
+
+/// Internal message that is used to load reply from the client.
+pub struct Receiver<T>(oneshot::Receiver<Result<T, core_client::Error>>);
+
+impl<T> Future for Receiver<T> {
+    type Item = T;
+    type Error = core_client::Error;
+
+    fn poll(&mut self) -> Poll<T, Self::Error> {
+        match self.0.poll() {
+            Ok(Async::Ready(Ok(x))) => Ok(Async::Ready(x)),
+            Ok(Async::Ready(Err(x))) => Err(x),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(e) => Err(core_client::Error::new(core_client::ErrorKind::Rpc, e)),
+        }
+    }
+}
+
+pub struct StreamOnce<K>(oneshot::Receiver<K>);
+
+impl<T> Stream for StreamOnce<T> {
+    type Item = T;
+    type Error = core_client::Error;
+
+    fn poll(&mut self) -> Poll<Option<T>, Self::Error> {
+        match self.0.poll() {
+            _ => unimplemented!(),
```

Comment:
What should this do?

Reply:
Oh, this guy was for the streams, but we don't have API that maps on the n-t-t API nicely, I'll see if I can implement `pull_blocks_to_tip`, then I'll fill this one.